### ISSUE TITLE
libjpeg-turbo should use calloc rather than malloc/memset

### ISSUE THREAD

**Issue by randomascii** [libjpeg-turbo/libjpeg-turbo#713]
**Have you searched the existing issues (both open and closed) in the libjpeg-turbo issue tracker to ensure that this bug report is not a duplicate?**

Yes

**Does this bug report describe one of the [two known and unsolvable issues with the JPEG format](https://libjpeg-turbo.org/pmwiki/uploads/About/TwoIssueswiththeJPEGStandard.pdf)?**

No

**Clear and concise description of the bug:**

libjpeg-turbo allocates large buffers by havling alloc_barray which calls alloc_large which calls jpeg_get_large which calls malloc. Later this memory is zeroed using jzero_far.

For large allocations in particular this is wasteful. Using calloc allows for greater efficiency. For large allocations calloc will obtain memory from the OS (VirtualAlloc or mmap) and this will come back from the OS zeroed. Since the C runtime knows whether calloc obtained memory from the OS or from the heap it knows whether or not a memset call is needed.

Since there is no way to obtain memory from the OS that has not been zeroed (for security reasons) it is wasteful to not use this facility when available, by calling calloc. This issue is discussed here: https://stackoverflow.com/questions/1538420/difference-between-malloc-and-calloc

**Steps to reproduce the bug (using *only* libjpeg-turbo):**

I haven't generated a stand-alone test case but this inefficiency is visible in the source code.

**Image(s) needed in order to reproduce the bug (if applicable):**

Any large .jpg image should suffice. It just needs to be large enough so that the array of JBLOCK structs is larger than the threshold for requesting memory from the OS. A 1024x1024 image should be large enough.

**Expected behavior:**

calloc should be used instead of malloc/memset.

**Observed behavior:**

malloc/memset are used.

**Platform(s) (compiler version, operating system version, CPU) on which the bug was observed:**

This was seen on Windows 10 with clang-cl, but it can most easily be observed in the source code:
https://github.com/libjpeg-turbo/libjpeg-turbo/blob/main/jmemnobs.c#L49

While jpeg_get_small _might_ not benefit from calloc (much), jpeg_get_large definitely would.

**libjpeg-turbo release(s), commit(s), or branch(es) in which the bug was observed (always test the tip of the main branch or the latest [stable pre-release](https://libjpeg-turbo.org/DeveloperInfo/PreReleases) to verify that the bug hasn't already been fixed):**

This issue is visible in tip-of-tree.

**If the bug is a regression, the specific commit that introduced the regression (use `git bisect` to determine this):**

Not a regression.

**Additional information:**

This isn't critical, but it does avoid a pass over all of the allocated memory. It also means that any memory that is not touched will never even be faulted in to the process, saving both CPU cycles and memory.

**Comment by dcommander** [comment]
This is a legacy of the design of libjpeg.  Bearing in mind that the libjpeg code base originated in 1991, it predates the widespread availability of ANSI/ISO C compilers.  Thus, the original implementation could not even rely on `malloc()`, much less `calloc()`.  It had different low-level memory manager modules for MS-DOS, Mac, temp files, and `malloc()`/`free()`.  libjpeg-turbo retained only the `malloc()`/`free()` memory manager, since the others were obsolete.  However, we are still beholden to the libjpeg design, which separates memory allocation (`jpeg_get_small()`, `jpeg_get_large()`) and zeroing (`jzero_far()`.)

There are few places in the libjpeg API library in which a buffer is allocated and then immediately zeroed, and those few places never account for more than 1-2k of memory each.

You assert that `jpeg_get_large()` would "definitely" benefit from `calloc()`, but you provide no evidence to support your claim.  You make a hand-waving statement that the issue can be "observed in the source code."  No, an issue can only be observed if an issue is observed.  If there is no observable negative behavior at run time, then there is no issue. 
Furthermore, `jpeg_get_large()` and `jpeg_get_small()` are public API functions, and their behavior is canonical.  That behavior cannot be changed without breaking backward compatibility with libjpeg.  Also, `jpeg_get_large()` is literally the same function as `jpeg_get_small()`.

The best we could do would be to leave `jpeg_get_*()` untouched and use `calloc()` only internally, for the rare instances in which a buffer is allocated and immediately zeroed.  But again, I didn't see any such instances that involved large buffers.

In other words, if you want me to do anything about this, then show me how the use of `calloc()` can lead to a visible performance improvement at run time.  I don't waste my valuable time fixing hypothetical issues.  I only fix real issues.

**Comment by randomascii** [comment]
Thank you for the response.

I had initially thought that this issue would happen with all .jpg images, but I have since realized that it is specific to progressive .jpg files. I hit it with a 10,000 x 8,000 progressive .jpg file that allocates and then zeroes (malloc/memset) three 160,000,000 buffers. This image can be found here:

https://drive.google.com/file/d/1Nw6YrxGURDryxgjnL1q8SlAG5Lc-ozGb/view?usp=drive_link

I can see that changing public API functions would be very difficult.

**Comment by dcommander** [comment]
OK, thanks. Since there is a concrete example, I will benchmark it.

**Comment by randomascii** [comment]
I am doing some benchmarking of this, so feel free to wait and I'll share some results.

The initial results suggest that the CPU/wall-clock time savings are minuscule.
However, I believe there is a secondary benefit on my test image. libjpeg-turbo allocates three 160,000,000 byte buffers for this image, but only uses two of them. With the calloc optimization that third buffer will never be faulted in, thus reducing the memory pressure. I will need to confirm this, as well as understand why I'm not seeing the expected CPU savings from this. It is possible that my test is not yet valid and that there will be savings once I do the change correctly.

**Comment by randomascii** [comment]
Because the buffers are zeroed just before being used the cost is small and I don't think it is worth changing anything. I apologize for not measuring this better before filing an issue - I made some incorrect assumptions.
